Fully revamp the release workflow, with PEP440 versions and publication triggered by a tag push

joanise opened the pull request:
This is currently "disabled" with the real actions being commented out or prefixed with "echo".

I'd like to test this by pushing an actual test tag, but I want code review before I do it, to make sure I correctly disabled the real work while keeping enough for us to see the result.

roedoejet commented on .github/workflows/pythonpublish.yml:
```diff
@@ -1,36 +1,139 @@
+      with:
+        name: python-package-distributions
+        path: dist/
+    - name: Publish distribution 📦 to TestPyPI
```
secrets.TEST_PYPI_API_TOKEN isn't available on g2p. I'll create one for secrets.PYPI_API_TOKEN but I don't think I have a test pypi account for making a PAT

dhdaines replied:
I have a test pypi account if you need someone to test this...

dhdaines replied:
In fact, I can just add g2p right now as a trusted publisher!

dhdaines replied:
If you want to change it to use trusted publishing, you can publish to the test.pypi.org project "gi2pi" (the name "g2p" was already taken).

Otherwise, I have an API token but I don't have permission to add it as a secret to this repository...

roedoejet replied:
oh nice - I didn't know you could do that. thanks @dhdaines !

joanise replied:
I agree, Aidan, we can drop test PyPI.
dhd, that would be great if we can use trusted publishing, thanks!

roedoejet reviewed (approved):
This is awesome @joanise - nicely done. I'll add the PAT for Pypi, but we'll also need to add a test pypi project and PAT if we want the test pypi job to run. I don't think it's necessary really though. we already know g2p works on pypi

joanise commented:
This workflow is now tested and fixed, and show work. There were a few bugs I discovered by pushing a fake v2.0.0a0 tag, which I have now deleted. I had the workflow set not to do (non-reversible) real work, just to be safe. But now it'll do the real thing when we push v2.0.0.
Please re-review @roedoejet @dhdaines to be on the safe side, but I'm fairly confident it's all good now.
